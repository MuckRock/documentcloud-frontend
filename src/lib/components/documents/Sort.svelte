<script lang="ts" context="module">
  export type SortField =
    | "score"
    | "title"
    | "created_at"
    | "updated_at"
    | "page_count";
  export type SortOrder = "asc" | "desc";

  export function isSortField(s?: string): s is SortField {
    if (!s) return false;
    return [
      "score",
      "title",
      "created_at",
      "updated_at",
      "page_count",
    ].includes(s);
  }
</script>

<script lang="ts">
  import { _ } from "svelte-i18n";
  import { ChevronDown16, SortAsc16, SortDesc16 } from "svelte-octicons";
  import Dropdown from "../common/Dropdown.svelte";
  import Menu from "../common/Menu.svelte";
  import SidebarItem from "../sidebar/SidebarItem.svelte";

  export let order: SortOrder;
  export let sort: SortField;
  export let fields: SortField[];
  export let query: string = "";
  export let onChange: (sort: SortField, order: SortOrder) => void = () => {};
</script>

<Dropdown>
  <SidebarItem slot="anchor">
    <div class="labelIcon" slot="start">
      {#if order === "asc"}
        <SortAsc16 />
      {:else}
        <SortDesc16 />
      {/if}
    </div>
    {$_("documentBrowser.sort.label")}
    <ChevronDown16 slot="end" />
  </SidebarItem>
  <Menu>
    <div class="field options">
      {#if query}
        <label class="field option" class:active={sort === "score"}>
          <input
            type="radio"
            bind:group={sort}
            value="score"
            on:change={() => onChange(sort, order)}
          />
          {$_(`documentBrowser.sort.fields.score`)}
        </label>
      {/if}
      {#each fields as f}
        <label class="field option" class:active={sort === f}>
          <input
            type="radio"
            bind:group={sort}
            value={f}
            on:change={() => onChange(sort, order)}
          />
          {$_(`documentBrowser.sort.fields.${f}`)}
        </label>
      {/each}
    </div>
    <div class="order options">
      <label class="order option desc" class:active={order === "desc"}>
        <SortDesc16 />
        <input
          type="radio"
          bind:group={order}
          value="desc"
          on:change={() => onChange(sort, order)}
        />
        {$_(`documentBrowser.sort.order.desc`)}
      </label>
      <label class="order option asc" class:active={order === "asc"}>
        <SortAsc16 />
        <input
          type="radio"
          bind:group={order}
          value="asc"
          on:change={() => onChange(sort, order)}
        />
        {$_(`documentBrowser.sort.order.asc`)}
      </label>
    </div>
  </Menu>
</Dropdown>

<style>
  .labelIcon {
    display: flex;
    align-items: center;
    justify-content: center;
  }

  .field.options {
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    gap: 0.5rem;
    padding: 0.5rem 0;
  }

  .order.options {
    display: flex;
    gap: 0;
    padding: 0 0.25rem 0.25rem;
  }

  .option {
    display: flex;
    align-items: center;
    gap: 0.5rem;
    cursor: pointer;
    padding: 0 0.5rem;

    font-family: "Source Sans Pro";
    font-size: var(--font-sm, 0.875rem);
    font-weight: var(--font-semibold, 600);
    font-style: normal;
    line-height: normal;
  }

  .option input {
    margin: 0;
  }

  .order.option {
    display: flex;
    padding: 0.375rem 0.5rem;
    justify-content: center;
    align-items: center;
    gap: 0.25rem;
    flex: 1 0 0;
    border: 1px solid var(--gray-3);
    background: var(--white);

    color: var(--gray-4, #5c717c);
    fill: var(--gray-3, #99a8b3);

    & input {
      display: none;
    }

    &:last-child {
      border-radius: 0rem 0.5rem 0.5rem 0rem;
      border-left: none;
    }

    &:first-child {
      border-radius: 0.5rem 0rem 0rem 0.5rem;
      border-right: none;
    }

    &.active {
      border: 1px solid var(--blue-4, #1367d0);
      color: var(--blue-4, #1367d0);
      fill: var(--blue-3, #4294f0);
      background: var(--blue-1, #eef3f9);
    }
  }
</style>
