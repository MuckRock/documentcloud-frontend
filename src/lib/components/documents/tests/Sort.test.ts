import { describe, it, expect } from "vitest";
import { render, fireEvent } from "@testing-library/svelte";
import Sort from "../Sort.svelte";
import type { SortField, SortOrder } from "../Sort.svelte";

describe("Sort.svelte", () => {
  const defaultProps = {
    order: "desc" as SortOrder,
    sort: "created_at" as SortField,
    fields: ["created_at", "title", "page_count"] as SortField[],
    query: "",
  };

  it("renders and opens menu for snapshot", async () => {
    const { container, getByText, getByLabelText } = render(Sort, defaultProps);

    // Click to open the menu
    const sortButton = getByText("Sort");
    await fireEvent.click(sortButton);

    // Verify all radio inputs are present
    expect(getByLabelText("Ascending")).toBeDefined();
    expect(getByLabelText("Descending")).toBeDefined();
    expect(getByLabelText("Date Created")).toBeDefined();
    expect(getByLabelText("Title")).toBeDefined();
    expect(getByLabelText("Total Pages")).toBeDefined();

    expect(container).toMatchSnapshot();
  });

  it("updates sort and order when selecting options", async () => {
    const { getByText, getByLabelText } = render(Sort, {
      ...defaultProps,
      order: "desc",
      sort: "created_at",
    });

    // Open the menu
    const sortButton = getByText("Sort");
    await fireEvent.click(sortButton);

    // Change sort field to "title"
    const titleOption = getByLabelText("Title");
    await fireEvent.click(titleOption);

    // Change order to ascending
    const ascOption = getByLabelText("Ascending");
    await fireEvent.click(ascOption);

    // Check if radio inputs are selected
    const titleInput = titleOption.parentNode?.querySelector("input");
    const ascInput = ascOption.parentNode?.querySelector("input");
    expect(titleInput?.checked).toBe(true);
    expect(ascInput?.checked).toBe(true);
  });
});
