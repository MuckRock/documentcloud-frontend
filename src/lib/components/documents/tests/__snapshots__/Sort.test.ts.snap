// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sort.svelte > renders and opens menu for snapshot 1`] = `
<div>
   
  <div
    class="anchor svelte-v01npp open"
    role="button"
    tabindex="0"
  >
    <span
      class="sidebarItem container svelte-1epk331"
      role="button"
      tabindex="0"
      title=""
    >
      <div
        class="labelIcon svelte-6mnr72"
        slot="start"
      >
        <svg
          class="octicon octicon-sort-desc"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 4.25a.75.75 0 0 1 .75-.75h7.5a.75.75 0 0 1 0 1.5H.75A.75.75 0 0 1 0 4.25Zm0 4a.75.75 0 0 1 .75-.75h4.5a.75.75 0 0 1 0 1.5H.75A.75.75 0 0 1 0 8.25Zm0 4a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5H.75a.75.75 0 0 1-.75-.75ZM13.5 10h2.25a.25.25 0 0 1 .177.427l-3 3a.25.25 0 0 1-.354 0l-3-3A.25.25 0 0 1 9.75 10H12V3.75a.75.75 0 0 1 1.5 0V10Z"
          />
        </svg>
      </div>
       
      <span
        class="label svelte-1epk331"
      >
        Sort
      </span>
       
      <svg
        class="octicon octicon-chevron-down"
        height="16"
        slot="end"
        viewBox="0 0 16 16"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.78 5.22a.749.749 0 0 1 0 1.06l-4.25 4.25a.749.749 0 0 1-1.06 0L3.22 6.28a.749.749 0 1 1 1.06-1.06L8 8.939l3.72-3.719a.749.749 0 0 1 1.06 0Z"
        />
      </svg>
    </span>
    
  </div>
   
  <div
    class="dropdown svelte-v01npp open"
    style="left: 0px; top: 0px; display: block;"
  >
    <div
      class="menu svelte-nrjt2v"
      role="menu"
    >
      <div
        class="field options svelte-6mnr72"
      >
         
        <label
          class="field option svelte-6mnr72 active"
        >
          <input
            class="svelte-6mnr72"
            type="radio"
            value="created_at"
          />
           
          Date Created
           
        </label>
        <label
          class="field option svelte-6mnr72"
        >
          <input
            class="svelte-6mnr72"
            type="radio"
            value="title"
          />
           
          Title
           
        </label>
        <label
          class="field option svelte-6mnr72"
        >
          <input
            class="svelte-6mnr72"
            type="radio"
            value="page_count"
          />
           
          Total Pages
           
        </label>
      </div>
       
      <div
        class="order options svelte-6mnr72"
      >
        <label
          class="order option desc svelte-6mnr72 active"
        >
          <svg
            class="octicon octicon-sort-desc"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 4.25a.75.75 0 0 1 .75-.75h7.5a.75.75 0 0 1 0 1.5H.75A.75.75 0 0 1 0 4.25Zm0 4a.75.75 0 0 1 .75-.75h4.5a.75.75 0 0 1 0 1.5H.75A.75.75 0 0 1 0 8.25Zm0 4a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5H.75a.75.75 0 0 1-.75-.75ZM13.5 10h2.25a.25.25 0 0 1 .177.427l-3 3a.25.25 0 0 1-.354 0l-3-3A.25.25 0 0 1 9.75 10H12V3.75a.75.75 0 0 1 1.5 0V10Z"
            />
          </svg>
           
          <input
            class="svelte-6mnr72"
            type="radio"
            value="desc"
          />
           
          Descending
        </label>
         
        <label
          class="order option asc svelte-6mnr72"
        >
          <svg
            class="octicon octicon-sort-asc"
            height="16"
            viewBox="0 0 16 16"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m12.927 2.573 3 3A.25.25 0 0 1 15.75 6H13.5v6.75a.75.75 0 0 1-1.5 0V6H9.75a.25.25 0 0 1-.177-.427l3-3a.25.25 0 0 1 .354 0ZM0 12.25a.75.75 0 0 1 .75-.75h7.5a.75.75 0 0 1 0 1.5H.75a.75.75 0 0 1-.75-.75Zm0-4a.75.75 0 0 1 .75-.75h4.5a.75.75 0 0 1 0 1.5H.75A.75.75 0 0 1 0 8.25Zm0-4a.75.75 0 0 1 .75-.75h2.5a.75.75 0 0 1 0 1.5H.75A.75.75 0 0 1 0 4.25Z"
            />
          </svg>
           
          <input
            class="svelte-6mnr72"
            type="radio"
            value="asc"
          />
           
          Ascending
        </label>
      </div>
    </div>
  </div>
</div>
`;
